### PR TITLE
OffsetDateProvider Class to reduce the likelihood of "request timestamp expired" exceptions.

### PR DESCRIPTION
**Pull request by PiezPiedPy**
Allows the client time to more closely match the GDAX API web servers time, reducing the likelihood of "request timestamp expired" exceptions due to small time discrepancies.

Depending on the time of day, I was receiving "request timestamp expired" exceptions when trying to communicate with the GDAX web server, after checking many times the synchronization of my clock with my local GMT time and finding no difference, I started troubleshooting and found that the problem was due to small time discrepancies between the GDAX web server and my local time, that would slowly increment throughout the day before dropping back down to zero and repeating the cycle.
The offset can get up to 90+ seconds and with the maximum offset of 30 seconds allowed by the GDAX web server I would start receiving the "request timestamp expired" exception.

Here is an example output of the offset when the exceptions started to appear, notice it is just above 30 seconds.
`GdaxTimeOffset: Local UTC Time   19/04/2018 5:52:45 pm +00:00`
`GdaxTimeOffset: GDAX Server Time 19/04/2018 5:53:16 pm +00:00`
`GdaxTimeOffset: Timestamp offset -31.593466758728`

The `OffsetDateProvider` Class I have created obtains the GDAX webserver time and calculates the difference from Local time and stores it for use by the `GdaxAuthenticationHandler` Class.
Time synchronization is not automatic and must be manually updated, it is refreshed by the user through the use of the method `RealOffset`.
If this method is never called by the user then the stored offset will remain at Zero and the `GdaxAuthenticationHandler` will operate as in previous releases. 

Since implementing this Time Offset class I have had no more issues with the "request timestamp expired" exception rearing its ugly head.

I have also enabled the building of the XML documentation file which is automatically included by Visual Studio into the NuGet package when Publishing is performed, allowing end users to view the summaries with Intelisense on Visual Studio.

### CODE REVIEW

**Review by rotiv123 — CHANGES_REQUESTED** [review]
Refactor your changes into a class that implements the IDateProvider interface.

**Comment by PiezPiedPy** [comment]
Is there a way the InnerHandler can be set in the GdaxApiClient class rather than the user having to set it.
I think it would be better to be able to use code such as:

`authenticator = new GdaxAuthenticationHandler(gdax_credentials, date_provider);`

rather than having to use:

`authenticator = new GdaxAuthenticationHandler(gdax_credentials, date_provider) { InnerHandler = new HttpClientHandler() };`

Must be a way but I cant see it, I will have a think tomorrow after some sleep 🛌  

Edit: Solved after some Pizza 😆

File: src/GdaxApi/GdaxApi/GdaxApi.csproj [hunk 1]
```diff
@@ -16,6 +16,16 @@
     <PackageReleaseNotes />
   </PropertyGroup>
 
+  <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Release|AnyCPU'">
+    <NoWarn>1701;1702;1705;1591</NoWarn>
```
Comment (rotiv123) [inline src/GdaxApi/GdaxApi/GdaxApi.csproj]: Why disable this warnings: 1701;1702;1705?

File: src/GdaxApi/GdaxApi/GdaxApiClient.cs [hunk 2]
```diff
@@ -26,6 +26,9 @@ public GdaxApiClient(GdaxAuthenticationHandler authenticationHandler, ISerialize
                 throw new ArgumentNullException(nameof(authenticationHandler));
             }
 
+            if (authenticationHandler.InnerHandler == null)
+                authenticationHandler.InnerHandler = new HttpClientHandler();
```
Comment (rotiv123) [inline src/GdaxApi/GdaxApi/GdaxApiClient.cs]: missing {}

File: src/GdaxApi/GdaxApi/Utils/OffsetDateProvider.cs [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+        public double UnixTimestamp => DateTime.UtcNow.ToUnixTimestamp() - StoredOffset;
+
+        /// <summary>
+        /// Returns the real time offset between the GDAX API web servers time and the clients local time in seconds.
+        /// <para>To allow the client time to more closely match the GDAX API web servers time reducing the likelyhood of a "request timestamp expired" exception,
+        /// you can refresh the stored time offset by using the optional parameter 'refresh'.</para>
+        /// <para>Note that the real vs stored offset can change over time, periodically check that the real vs stored value does not exceed
+        /// 30 seconds and refresh when it does, or you will start to recieve "request timestamp expired" exceptions.</para>
+        /// </summary>
+        public async Task<double> RealOffset(GdaxApiClient api, bool refresh = false)
```
Comment (rotiv123) [inline src/GdaxApi/GdaxApi/Utils/OffsetDateProvider.cs]: Whats the point of calling this method to get the API server time?
This method should receive directly an "ApiTime" and update its internal offset. That way there is no need to use the "refresh" parameter.

File: src/GdaxApi/GdaxApi/Utils/OffsetDateProvider.cs [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+        {
+            ApiTime exchangetime;
+
+            try
+            {
+                exchangetime = await api.GetTime().SendAsync();
+            }
+            catch { throw; }
```
Comment (rotiv123) [inline src/GdaxApi/GdaxApi/Utils/OffsetDateProvider.cs]: No useless catch statements please

File: src/GdaxApi/GdaxApi/Utils/OffsetDateProvider.cs [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    {
+        /// <summary>
+        /// Gets the stored time offset between the GDAX API web servers time and the clients local time in seconds.
+        /// <para>Note that the stored vs real offset can change over time, periodically check that the stored vs real value does not exceed 30 seconds
+        /// or you will start to recieve "request timestamp expired" exceptions.</para>
+        /// <para>Use the <see cref="RealOffset"/> method with the optional parameter 'refresh' set to true to refresh the stored offset.</para>
+        /// </summary>
+        public double StoredOffset { get; private set; } = 0d;
+
+        public DateTimeOffset Now => DateTimeOffset.Now;
```
Comment (rotiv123) [inline src/GdaxApi/GdaxApi/Utils/OffsetDateProvider.cs]: This property must represent the same date as the "UnixTimestamp" property, apart from the timezone offset. You forgot to subtract the "StoredOffset"

**Review by rotiv123 — CHANGES_REQUESTED** [review]

File: src/GdaxApi/GdaxApi/GdaxApi.csproj (hunk 1 above)
Comment (PiezPiedPy) [reply src/GdaxApi/GdaxApi/GdaxApi.csproj]: They where previously defined, I did not add them. I only added the 1591 warning.
Should I remove them?

File: src/GdaxApi/GdaxApi/GdaxApiClient.cs (hunk 2 above)
Comment (PiezPiedPy) [reply src/GdaxApi/GdaxApi/GdaxApiClient.cs]: Didn't know they where required, works in testing, I'll add them.
Just a coding habit of mine not to use them for single lines

Edit: Done

File: src/GdaxApi/GdaxApi/Utils/OffsetDateProvider.cs (hunk 5 above)
Comment (PiezPiedPy) [reply src/GdaxApi/GdaxApi/Utils/OffsetDateProvider.cs]: Ok, I'll add in the offset

Edit: Done ;)

File: src/GdaxApi/GdaxApi/Utils/OffsetDateProvider.cs (hunk 3 above)
Comment (PiezPiedPy) [reply src/GdaxApi/GdaxApi/Utils/OffsetDateProvider.cs]: The method returns the actual Offset not the stored offset for the user to use for checking how large the offset is.
Idea is to save calling the GDAX webserver repeatedly and only use the refresh when needed. i.e. only refresh when the time offset gets close to 30 seconds. 
I could add a bool property so the end user can set the class to an automatic or manual refresh mode.
btw, do you think automatic if implemented, refreshes the offset before each call to the GDAX webserver or use a timer that waits till the offset gets close to 30 seconds before refreshing?

I will add this method:
`        public void Refresh(ApiTime exchangetime)
        {
            this.StoredOffset = DateTime.UtcNow.ToUnixTimestamp() - exchangetime.UnixTimestamp;
        }
`

Edit: Changed the original method and added the Refresh method similar as above.
No need to mess about with any automatic stuff as the user can implement that if needed.

File: src/GdaxApi/GdaxApi/Utils/OffsetDateProvider.cs (hunk 4 above)
Comment (PiezPiedPy) [reply src/GdaxApi/GdaxApi/Utils/OffsetDateProvider.cs]: I was throwing the exception back to the user to handle.
Any idea what exception I should throw, or should I remove the try catch all together? (removing seems a bad idea to me)

Edit: removed this method now, disregard the above.

File: src/GdaxApi/GdaxApi/GdaxApi.csproj (hunk 1 above)
Comment (rotiv123) [reply src/GdaxApi/GdaxApi/GdaxApi.csproj]: yes

**Review by rotiv123 — CHANGES_REQUESTED** [review]

File: src/GdaxApi/GdaxApi/GdaxApi.csproj (hunk 1 above)
Comment (PiezPiedPy) [reply src/GdaxApi/GdaxApi/GdaxApi.csproj]: Done 😉

**Pull request merged**